### PR TITLE
[debugger] Work around a debugger issue when using the interpreter

### PR DESCRIPTION
**Pull request by thaystg**
TODO:
Only start runtime using mono_main when it's on interpreter mode and debug enabled.

CONS:
With this implementation we see in the log a message like this:
```
Usage is: mono [options] program [program-options]

Development:
    --aot[=<options>]      Compiles the assembly to native code
    --debug=ignore         Disable debugging support (on by default)
    --debug=[<options>]    Disable debugging support or enable debugging extras, use --help-debug for details
    --debugger-agent=options Enable the debugger agent
    --profile[=profiler]   Runs in profiling mode with the specified profiler module
    --trace[=EXPR]         Enable tracing, use --help-trace for details
    --help-devel           Shows more options available to developers

Runtime:
    --config FILE          Loads FILE as the Mono config
    --verbose, -v          Increases the verbosity level
    --help, -h             Show usage information
    --version
```

Maybe we could pass on args --version then we would see the mono version but wouldn't see this help message.

### CODE REVIEW

File: runtime/monovm-bridge.m [hunk 1]
```diff
@@ -85,7 +82,16 @@
...
 	mono_install_ftnptr_eh_callback (xamarin_ftnptr_exception_handler);
 
-	mono_jit_init_version ("MonoTouch", "mobile");
+	const char *argc[] = {
+		"",
+		"--interp=-all"
+	};
+
+	if (xamarin_init_mono_debug)
+		mono_main (2, (char**)argc);
```
Comment (rolfbjarne) [inline runtime/monovm-bridge.m:91]: Could you explain runtime-wise why this works around the problem?

**Comment by thaystg** [comment]
Yes, sure.
The issue on runtime side was that we are initializing interpreter before setting the debugger attributes. So when we execute this code: https://github.com/dotnet/runtime/blob/5529dc79e9f35d70e9d01318d31844984b860dd4/src/mono/mono/mini/interp/interp.c#L8121
mdb_optimizations was false, then we keep the default optimizations of the interpreter which was breaking the debugger experience.
In this workaround I'm passing `--interp=-all` in the parameters, so I'm disabling the interpreter optimizations and the debugger works well.

File: runtime/monovm-bridge.m (hunk 1 above)
Comment (thaystg) [reply runtime/monovm-bridge.m:91]: Yes, sure.
The issue on runtime side was that we are initializing interpreter before setting the debugger attributes. So when we execute this code: https://github.com/dotnet/runtime/blob/5529dc79e9f35d70e9d01318d31844984b860dd4/src/mono/mono/mini/interp/interp.c#L8121
mdb_optimizations was false, then we keep the default optimizations of the interpreter which was breaking the debugger experience.
In this workaround I'm passing `--interp=-all` in the parameters, so I'm disabling the interpreter optimizations and the debugger works well.

Comment (thaystg) [reply runtime/monovm-bridge.m:91]: I'm almost sure that it's a problem only on .net 6, but I would ask someone to test to double check. It's really easy to reproduce:

	```
public static void MethodMyMethod()
	{
		var e2 = new Exception("hi thays");
		var e1 = e2;
		if (e1.Equals(e2))
			Console.WriteLine("equals");
	}
```
Adding a breakpoint on if statement and looking at locals, with the optimizations turned on we only see the content of variable e2, e1 is null.

**Comment by mandel-macaque** [comment]
the xtra failures are interesting, is complaining due to a binding problem, when it should not happen. Merge with main, if issues are the same I'll fix the PR or main

File: runtime/monovm-bridge.m [hunk 2]
```diff
@@ -85,7 +82,36 @@
...
+		// to get the side effects).
+		fflush (stdout);
+		int originalStdout = dup (STDOUT_FILENO);
+		int devnull = open ("/dev/null", O_RDWR);
+		dup2 (devnull, STDOUT_FILENO);
+		mono_main (2, (char**)argc);
+		fflush (stdout);
+		dup2 (originalStdout, STDOUT_FILENO);
+	} else {
+		mono_jit_init_version ("MonoTouch", "mobile");
```
Comment (thaystg) [inline runtime/monovm-bridge.m:112]: I think we need this code here:
	if (xamarin_init_mono_debug)
		mono_debug_init (MONO_DEBUG_FORMAT_MONO);

Comment (rolfbjarne) [reply runtime/monovm-bridge.m:112]: Ah yeah

Comment (rolfbjarne) [reply runtime/monovm-bridge.m:115]: Fixed

**Comment by rolfbjarne** [comment]
Mac Catalyst apps crash with like this with this fix: https://gist.github.com/rolfbjarne/293ecffb7fcf3e4169494b81d7e4cd97

**Comment by thaystg** [comment]
> Mac Catalyst apps crash with like this with this fix: https://gist.github.com/rolfbjarne/293ecffb7fcf3e4169494b81d7e4cd97

This is exactly the scenario that I was testing and I don't see this crash:
I'm building xamarin-macios locally and then:
`cp ./_build/Microsoft.MacCatalyst.Runtime.maccatalyst-x64/runtimes/maccatalyst-x64/native/libxamarin-dotnet-debug.* /Users/thaysgrazia/testJason/bin/Debug/net6.0-maccatalyst/maccatalyst-x64/testJason.app/Contents/MonoBundle/`

Then I debug the app and I can see the locals variables correctly.

**Comment by rolfbjarne** [comment]
> > Mac Catalyst apps crash with like this with this fix: [gist.github.com/rolfbjarne/293ecffb7fcf3e4169494b81d7e4cd97](https://gist.github.com/rolfbjarne/293ecffb7fcf3e4169494b81d7e4cd97)
> 
> This is exactly the scenario that I was testing and I don't see this crash: I'm building xamarin-macios locally and then: `cp ./_build/Microsoft.MacCatalyst.Runtime.maccatalyst-x64/runtimes/maccatalyst-x64/native/libxamarin-dotnet-debug.* /Users/thaysgrazia/testJason/bin/Debug/net6.0-maccatalyst/maccatalyst-x64/testJason.app/Contents/MonoBundle/`
> 
> Then I debug the app and I can see the locals variables correctly.

You need this commit for it to crash reliably:

https://github.com/xamarin/xamarin-macios/pull/15451/commits/9b78d773672a277da586d6eaf30aa321b70c0462

Otherwise it only crashes on macOS 10 and macOS 11, on newer versions you get random behavior (which happens to work).

**Comment by rolfbjarne** [comment]
Test failures are unrelated (https://github.com/xamarin/maccore/issues/2595).

**Review by rolfbjarne — APPROVED** [review]

File: runtime/monovm-bridge.m [hunk 3]
```diff
@@ -82,12 +82,18 @@
 	mono_install_unhandled_exception_hook (xamarin_unhandled_exception_handler, NULL);
 	mono_install_ftnptr_eh_callback (xamarin_ftnptr_exception_handler);
 
+#if DOTNET
+	bool use_mono_workaround = xamarin_init_mono_debug;
+#else
+	bool use_mono_workaround = false;
```
[the remark this answers is not in the document]
Comment (rolfbjarne) [reply runtime/monovm-bridge.m]: Just to follow up: legacy scenarios will be fixed with https://github.com/xamarin/xamarin-macios/pull/15507.

**Pull request merged**